platform/packages/versioning/src/release.rs

```diff
@@ -1,279 +1,268 @@
+
+        VERSION_STORAGE_KEY
+            .load(storage)
+            .inspect(|_| VERSION_STORAGE_KEY.remove(storage))
+            .map(|code| Self::instance(Id::prev(), code))
+    }
+
+    fn instance(id: Id, code: Version) -> Self {
+        Self { id, code }
     }
 
-    pub fn allow_software_update(&self, current: &Version, new: &Version) -> Result<(), StdError> {
-        check_storage_match(
-            current.storage,
-            new.storage,
-            "the new software storage version",
-        )?;
+    pub fn update_software(self, to: Self) -> Result<Self, StdError> {
+        self.check_storage_match(to.code)
+            .and_then(|()| self.allow_software_update_int(to))
+    }
 
-        self.allow_software_update_int(current, new)
+    pub fn update_software_and_storage(self, to: Self) -> Result<Self, StdError> {
+        self.check_storage_adjacent(to.code)
+            .and_then(|()| self.allow_software_update_int(to))
     }
 
-    pub fn allow_software_and_storage_update<const FROM_STORAGE_VERSION: VersionSegment>(
-        &self,
-        current: &Version,
-        new: &Version,
-    ) -> Result<(), StdError> {
-        check_storage_match(
-            current.storage,
-            FROM_STORAGE_VERSION,
-            "the expected origin storage version",
-        )?;
-
-        if current.storage.wrapping_add(1) == new.storage {
-            self.allow_software_update_int(current, new)
+    fn allow_software_update_int(&self, new: Self) -> Result<Self, StdError> {
+        let current = self.code;
+        if current < new.code || (self.id == Id::dev() && current == new.code) {
+            Ok(new)
         } else {
             Err(StdError::generic_err(
-                "The storage version is not adjacent to the current one!",
+                "The software version does not increase monotonically!",
             ))
         }
     }
 
-    fn allow_software_update_int(&self, current: &Version, new: &Version) -> Result<(), StdError> {
-        if current.software < new.software
-            || (self.0 == Self::DEV_RELEASE && current.software == new.software)
-        {
+    fn check_storage_match(&self, other: Version) -> Result<(), StdError> {
+        if self.code.same_storage(&other) {
             Ok(())
         } else {
-            Err(StdError::generic_err(
-                "The software version does not increase monotonically!",
-            ))
+            Err(StdError::generic_err(format!(
+                "The storage versions do not match! The new software version is \"{other}\"!",
+            )))
+        }
+    }
+
+    fn check_storage_adjacent(&self, next: Version) -> Result<(), StdError> {
+        if self.code.next_storage(&next) {
+            Ok(())
+        } else {
+            Err(StdError::generic_err(format!(
+                "The new version \"{next}\" is not adjacent to the current one \"{current}\"!",
+                current = self.code
```

Comment:
[Minor] I would argue this would have mattered more if there was more than one value, but I am not willing to die on that hill, so it's up to your opinion on the topic.